File: app/src/commands/plugins.rs
//! Plugin management IPC commands.
//!
//! These commands let the UI inspect, load, and unload plugins. Hot-reload is
//! handled transparently by the watcher background task; `plugin_reload` is
//! provided for cases where the user wants to force a reload without waiting
//! for a file event.

use tauri::State;

use pixhaus_plugins::PluginInfo;

use crate::error::{AppCommandError, CommandResult};
use crate::state::AppState;

/// Lists all currently loaded plugins and their metadata.
#[tauri::command(async, rename_all = "snake_case")]
pub async fn plugin_list(state: State<'_, AppState>) -> CommandResult<Vec<PluginInfo>> {
    Ok(state.plugins.list())
}

/// Triggers a re-scan of the plugin directory and loads any new plugins.
///
/// This is the manual equivalent of restarting the editor. Plugins already
/// loaded are NOT reloaded; use `plugin_reload` for that.
#[tauri::command(async, rename_all = "snake_case")]
pub async fn plugin_scan(state: State<'_, AppState>) -> CommandResult<usize> {
    state
        .plugins
        .scan()
        .await
        .map_err(|e| AppCommandError::Validation {
            detail: e.to_string(),
        })
}

/// Unloads the named plugin and reloads it from disk.
///
/// Returns an error if the plugin is not currently loaded or if the reload
/// fails (e.g. manifest parse error, WASM load failure). On failure the plugin
/// stays unloaded — the caller should surface the error to the user.
#[tauri::command(async, rename_all = "snake_case")]
pub async fn plugin_reload(name: String, state: State<'_, AppState>) -> CommandResult<()> {
    state
        .plugins
        .reload(&name)
        .await
        .map_err(|e| AppCommandError::Validation {
            detail: e.to_string(),
        })
}

/// Unloads the named plugin. Its registered verbs, tools, and commands are
/// removed immediately.
#[tauri::command(async, rename_all = "snake_case")]
pub async fn plugin_unload(name: String, state: State<'_, AppState>) -> CommandResult<()> {
    state
        .plugins
        .unload(&name)
        .map_err(|e| AppCommandError::Validation {
            detail: e.to_string(),
        })
}

#[cfg(test)]
mod tests {
    use super::*;

    // Smoke-test that PluginInfo is serialisable. The IPC bridge requires all
    // return types to implement serde::Serialize.
    #[test]
    fn plugin_info_serialises() {
        let info = PluginInfo {
            name: "echo".into(),
            version: "0.1.0".into(),
            description: Some("Echo verb for testing".into()),
            author: None,
            dir: "/home/user/.pixhaus/plugins/echo".into(),
            verb_count: 1,
        };
        let json = serde_json::to_string(&info).unwrap();
        assert!(json.contains("echo"));
    }
}

File: ui/src/lib/commands/plugins.ts
// Plugin management commands.
//
// Mirrors the Rust `PluginInfo` shape from `plugins/src/registry.rs`.
// `description` and `author` are `Option<String>` with `skip_serializing_if`,
// so they may be absent on the wire.

import { invoke } from "../ipc";

// ── response types ────────────────────────────────────────────────────────────

export interface PluginInfo {
  /** Plugin name from the manifest. */
  name: string;
  /** Plugin version from the manifest. */
  version: string;
  /** Optional description from the manifest. */
  description?: string;
  /** Optional author from the manifest. */
  author?: string;
  /** Absolute path to the plugin directory. */
  dir: string;
  /** Number of verbs this plugin registered. */
  verb_count: number;
}

// ── commands ──────────────────────────────────────────────────────────────────

/** Returns all currently loaded plugins. */
export function pluginList(): Promise<PluginInfo[]> {
  return invoke<PluginInfo[]>("plugin_list");
}

/**
 * Re-scans the plugin directory and loads any newly added plugins.
 * Returns the count of plugins now loaded. Does NOT reload plugins
 * that were already loaded — use {@link pluginReload} for that.
 */
export function pluginScan(): Promise<number> {
  return invoke<number>("plugin_scan");
}

/** Unloads and reloads the named plugin from disk. */
export function pluginReload(name: string): Promise<void> {
  return invoke<void>("plugin_reload", { name });
}

/** Unloads the named plugin. Its verbs are deregistered immediately. */
export function pluginUnload(name: string): Promise<void> {
  return invoke<void>("plugin_unload", { name });
}
